python/cuml/cuml/prims/label/classlabels.py

```diff
@@ -50,27 +50,41 @@
 
 
 validate_kernel_str = r"""
-({0} *x, int x_n, {0} *labels, int n_labels, int *out) {
+({0} *labels, int n_labels,
+ {0} *classes, int n_classes,
+ int n_passes, int pass_size,
+ int *out) {
   int tid = blockDim.x * blockIdx.x + threadIdx.x;
+  bool found = false;
 
-  extern __shared__ {0} label_cache[];
-  for(int i = threadIdx.x; i < n_labels; i+=blockDim.x)
-    label_cache[i] = labels[i];
-
-  if(tid >= x_n) return;
-
-  __syncthreads();
+  extern __shared__ {0} class_cache[];
+
+  int unmapped_class;
+  if (tid < n_labels)
+    unmapped_class = labels[tid];
+  for (int pass = 0; pass < n_passes; pass++) {
+    int offset = pass * pass_size;
+    int to_analyze = min(pass_size, n_classes - offset);
+    for (int i = threadIdx.x; i < to_analyze; i+=blockDim.x)
+      class_cache[i] = classes[offset + i];
+
+    __syncthreads();
+
+    if (!found && tid < n_labels) {
```

Comment:
found will always be false at this point?

Reply:
A thread with `tid < n_labels` does not check the presence of a class for a label, but it is still necessary to correctly fetch the data for every pass. The `found` value will indeed always be false, but a thread with `tid < n_labels` will never write the final output, it should thus be safe.